### ISSUE TITLE
The dataset for SUNSET-forecast

### ISSUE THREAD

**Issue by yuyyuu** [yuhao-nie/Stanford-solar-forecasting-dataset#3]
Thanks for your great work. However, I have some trobule when I try to reproduce the SUNSET-forecast model. I cann't find the file 'forecast_dataset.h5py' for this model. I also cannot find the files 'all_times_highfreq.npy' 'all_images_highfreq.npy' and 'pv_output_valid.pkl'. How can I get these files?

**Comment by yuhao-nie** [comment]
In the [benchmark dataset](https://purl.stanford.edu/dj417rh1007), we named it differently as "2017_2019_images_pv_processed.hdf5" for ease of recognizing the data.  The files 'all_times_highfreq.npy' 'all_images_highfreq.npy' and 'pv_output_valid.pkl' are outputs from the data processing notebooks. You will need to run the data processing notebooks using either our raw dataset or your own raw dataset to get these data files. The benchmark dataset are already processed and can be directly used. For details please check our dataset paper.

**Comment by yuhao-nie** [comment]
> In the [benchmark dataset](https://purl.stanford.edu/dj417rh1007), we named it differently as "2017_2019_images_pv_processed.hdf5" for ease of recognizing the data. The files 'all_times_highfreq.npy' 'all_images_highfreq.npy' and 'pv_output_valid.pkl' are outputs from the data processing notebooks. You will need to run the data processing notebooks using either our raw dataset or your own raw dataset to get these data files. The benchmark dataset are already processed and can be directly used. For details please check our dataset paper.

Apology. The '2017_2019_images_pv_processed.hdf5' file contains essentially the images and PV pairs for each time stamp, which can be directly used for SUNSET-nowcast model, but not for SUNSET-forecast model. You need to run the data_preprocess_forecast notebook with your parameters, e.g., forecasting horizon, sampling interval, length of history, etc. to obtain the 'forecast_dataset.hdf5' as input for SUNSET-forecast. Time stamps associated with '2017_2019_images_pv_processed.hdf5' are provided separately in the [benchmark dataset](https://purl.stanford.edu/dj417rh1007).

**Comment by yuyyuu** [comment]
Thanks for your answer. That's helpful. But I encountered another problem when i trained the SUNSET-nowcast model. The loss of training set and validation set dones't decrease, and the error of the trained model is particularly large. All predictions on the test set are almost constant values. I don't make any changes other than changing the dataset's folder.
![loss](https://user-images.githubusercontent.com/77527457/187577529-addacd14-31a7-4825-aec3-7bd262fb200d.PNG)

**Comment by guuanrui** [comment]
> 感谢您的出色工作。但是，当我尝试重现日落预报模型时，我遇到了一些麻烦。我找不到此模型的文件“forecast_dataset.h5py”。我也找不到文件“all_times_highfreq.npy”all_images_highfreq.npy“和”pv_output_valid.pkl”。如何获取这些文件？
hi


> 感谢您的出色工作。但是，当我尝试重现日落预报模型时，我遇到了一些麻烦。我找不到此模型的文件“forecast_dataset.h5py”。我也找不到文件“all_times_highfreq.npy”all_images_highfreq.npy“和”pv_output_valid.pkl”。如何获取这些文件？

hi, can you tell me how do you get these 

> 

hi, can you tell me how do you get these files? thank you.

**Comment by guuanrui** [comment]
> Thanks for your great work. However, I have some trobule when I try to reproduce the SUNSET-forecast model. I cann't find the file 'forecast_dataset.h5py' for this model. I also cannot find the files 'all_times_highfreq.npy' 'all_images_highfreq.npy' and 'pv_output_valid.pkl'. How can I get these files?

hi, can you tell me how do you get these files? thank you.

**Comment by nilsleh** [comment]
I am also trying to generate the forecast dataset from the published `2017_2019_images_pv_processed.hdf5` and `times_test.npy` and `times_trainval.npy` files, so these are the changes I added to the notebook to create those arrays.

```
def create_samples(pv_data, all_images, all_times):
    """Create the desired forecast samples.
    
    Args:
        pv_data: numpy array of the pv values 
        all_images: hdf5 dataset with the images
        all_times: the desired time steps to consider from training or testing
    """
    # Create forecast training data file
    n_images = all_times.shape[0]
    for sampling_interval in sampling_interval_all:
        # Initialize variables to save pv values
        image_log = np.zeros([n_images,stack_height+1]+output_img_shape,dtype = 'uint8')
        pv_log = np.zeros((n_images,stack_height+1))
        pv_pred = np.zeros(n_images)
        validity_mask = np.ones(n_images,dtype = bool)
        tic = time.process_time()
        last_valid_index = 0

        sampling_interval_td = datetime.timedelta(minutes = sampling_interval) - datetime.timedelta(seconds=1)
        for i in range(0,n_images):

            # See if the specified sampling frequency is met 
            if all_times[i] - all_times[last_valid_index] > sampling_interval_td:

                # Collecting groud truth for predicted value
                pred_time = all_times[i]+datetime.timedelta(minutes=forecast_horizon)
                
                pv_pred_idx = find_time_within_nparray(all_times, pred_time)
                if pv_pred_idx is None:# if prediction ground truth not found
                    validity_mask[i] = False
                    print(all_times[i],'has no PV pred')
                else: 
                    pv_pred[i] = pv_data[pv_pred_idx] 

                # Collecting image log and PV log
                for j in range(stack_height+1):
                    log_time = all_times[i] - datetime.timedelta(minutes = j)
                    # Collecting a stack of image
                    log_time_idx = find_time_within_nparray(all_times,log_time)
                    if log_time_idx is not None:
                        image_log[i,j] = all_images[log_time_idx]
                    else:
                        validity_mask[i] = False
                        print(all_times[i],'has no image log')
                        break

                    # Collecting a stack of PV value
                    pv_log_idx = find_time_within_nparray(all_times, log_time)
                    # Check if PV value present
                    if pv_log_idx is None:
                        validity_mask[i] = False
                        print(all_times[i],'has no PV log')
                        break
                    else: 
                        pv_log[i,j] = pv_data[pv_log_idx]    

            else: # if this is in between the sampling points, discard
                validity_mask[i] = False
            
            if validity_mask[i]:
                last_valid_index = i
            
            # Prompt progress of current work
            if i%100 == 0:
                print('processed {0}/{1} images'.format(i,len(all_times)))
                if i%1000 == 0 and i>0:
                    print('For sampling frequency: ',sampling_interval,' minutes')
                    print('Expected finishing time:', datetime.datetime.now()+
                            datetime.timedelta(seconds = (time.process_time() - tic)*(len(all_times)/i-1)))
        
        # Only pick out the valid time points
        all_times = all_times[validity_mask]
        image_log = image_log[validity_mask]
        pv_log = pv_log[validity_mask]
        pv_pred = pv_pred[validity_mask]

    return all_times, image_log, pv_log, pv_pred
```

For training/validation set:
```
train_val_times = np.load("times_trainval.npy", allow_pickle=True)
with h5py.File("2017_2019_images_pv_processed.hdf5", "r") as f:
    train_val_images = f["trainval"]["images_log"]
    train_val_pv = f["trainval"]["pv_log"]

    train_all_times, train_image_log, train_pv_log, train_pv_pred = create_samples(train_val_pv[:], train_val_images, train_val_times)
```

For test set:
```
test_times = np.load("times_test.npy", allow_pickle=True)
with h5py.File("2017_2019_images_pv_processed.hdf5", "r") as f:
    test_images = f["test"]["images_log"]
    test_pv = f["test"]["pv_log"]

    test_all_times, test_image_log, test_pv_log, test_pv_pred = create_samples(test_pv[:], test_images, test_times)
```

**Comment by yuhao-nie** [comment]
Thanks @nilsleh for the clarification. I think that's correct if you want to quickly get some forecast data. I just want to make some notes here regarding the issues I recently noticed with the above approach: in the original data_preprocess_forecast.ipynb notebook, we used a higher frequency PV data files “pv_output_valid.pkl”, which can be obtained by running the data_preprocess_PV.ipynb with the raw PV data (can be accessed from the link provided in the GitHub repo). The PV data in the benchmark dataset are essentially a subset of this higher-frequency one (we are trying to pair images with PV logs to form the benchmark, so the ones end up without finding a pair are thus filtered out.). You will end up with a slightly smaller forecast dataset if you use the PV data from the benchmark data to form the forecast dataset. I would appreciate if anyone have used the forecast dataset derived from the benchmark dataset to train SUNSET forecast model to provide some feedbacks on the model performance. How much impact does the data make if you compared to the results reported in the SKIPP'D paper?

Update: Be careful! the time stamps in times_test.npy are not sorted (from small to large), that would cause issue when using the above code, specifically with the current find_time_within_nparray() function, which assumes that the time stamps are sorted. So use the following code for the test set:
`sorted_indices = np.argsort(test_times)`
`test_times = test_times[sorted_indices]`
`test_pv = test_pv[sorted_indices]`
`test_images = test_images[sorted_indices]`
`create_samples(test_pv[:], test_images, test_times)`

**Comment by yuhao-nie** [comment]
@yuyyuu have you solved this problem?

**Comment by doitslow** [comment]
Your preprocess notebook indicates that there are 1904 test samples generated for forecasting task, but SUNSET_forecast.ipynb did testing on 12248 samples. What's the gap?
